File: 0200-number-of-islands/0200-number-of-islands.py
class Solution(object):
    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        num_of_islands = 0
        
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == '0':
                    continue
                else:
                    self.dfs(grid, i, j)
                    num_of_islands+=1
        
        return num_of_islands
    
    def dfs(self, grid, row, col):
        if (row < 0 or col < 0 or row >= len(grid) or col >= len(grid[0]) or grid[row][col] == '0'):
            return
        else:
            grid[row][col] = '0'
            self.dfs(grid, row-1, col)
            self.dfs(grid, row+1, col)
            self.dfs(grid, row, col-1)
            self.dfs(grid, row, col+1)
        

File: 0200-number-of-islands/0200-number-of-islands.java
class Solution {
    public int numIslands(char[][] grid) {
        int count = 0;

        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                if (grid[i][j] == '1') {
                    count += 1;
                    explore(grid, i, j);
                }
            }
        }

        return count;
    }

    public void explore(char[][] grid, int i, int j) {
        if (i < 0 || i >= grid.length || j < 0 || j >= grid[0].length || grid[i][j] == '0') {
            return;
        }

        grid[i][j] = '0';

        explore(grid, i-1, j);
        explore(grid, i+1, j);
        explore(grid, i, j-1);
        explore(grid, i, j+1);
    }
}